### PR TITLE
[9/n] tlsmux: gRPC inbound integration

### PR DESCRIPTION
**Pull request by jronak**
gRPC transport exposes a new inbound option, `InboundMuxTLS` - this enables tlsmux listener with the provided tls configuration. To receive tlsmux metrics, users can provide metric instance with the `Meter` transport option.

### CODE REVIEW

File: transport/grpc/options.go [hunk 1]
```diff
@@ -90,6 +92,15 @@ func Logger(logger *zap.Logger) TransportOption {
 	}
 }
 
+// Meter sets a meter to use for transport metrics.
+//
+// The default is to not emit any metrics.
```
Comment (gopalchouhan) [inline transport/grpc/options.go:97]: This looks to be a setter option method, is this the right place for this comment?

File: transport/internal/tlsmux/listener.go [hunk 2]
```diff
@@ -21,45 +21,130 @@
...
+
+	_sniffReadTimeout    = time.Second * 10
+	_tlsHandshakeTimeout = time.Second * 10
+)
+
+// Config describes how listener should be configured.
+type Config struct {
+	Listener  net.Listener
+	TLSConfig *tls.Config
+
```
Comment (gopalchouhan) [inline transport/internal/tlsmux/listener.go:46]: nit: I don't think we need a new line here. Can you please remove this?

File: transport/internal/tlsmux/listener.go [hunk 3]
```diff
@@ -21,45 +21,130 @@
...
-		}
-		return c, err
+// serveConnection muxes the given connection and sends muxed connection to the
+// connection channel.
+func (l *listener) serveConnection(conn net.Conn, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	c, err := l.mux(conn)
+	if err != nil {
+		conn.Close()
```
Comment (gopalchouhan) [inline transport/internal/tlsmux/listener.go:140]: Any particular reason to discard the err here without logging/passing?

File: transport/internal/tlsmux/listener.go [hunk 4]
```diff
@@ -21,45 +21,130 @@
...
+
+	"go.uber.org/net/metrics"
+	"go.uber.org/zap"
 )
 
+var (
+	errListenerClosed = errors.New("listener closed")
+
+	_sniffReadTimeout    = time.Second * 10
+	_tlsHandshakeTimeout = time.Second * 10
```
Comment (gopalchouhan) [inline transport/internal/tlsmux/listener.go:39]: curious to know on how did we come up with this timeout value?

File: transport/internal/tlsmux/listener_test.go [hunk 5]
```diff
@@ -148,6 +203,103 @@ func TestMux(t *testing.T) {
...
+
+	for i := 0; i < totalConnections; i++ {
+		conn, err := muxLis.Accept()
+		require.NoError(t, err)
+
+		wg.Add(1)
+		go func(c net.Conn) {
+			defer wg.Done()
+
+			buf := make([]byte, len(msg))
```
Comment (gopalchouhan) [inline transport/internal/tlsmux/listener_test.go:293]: looks like both tls and plain client are writing same message. Should we write different msg to both connection to ensure there are being processed by right set of routines?

Comment (jronak) [reply transport/internal/tlsmux/listener_test.go]: Sorry, this wasn't rebased during the time of review. I'll address this in a separate pull request

File: transport/internal/tlsmux/listener.go (hunk 4 above)
Comment (jronak) [reply transport/internal/tlsmux/listener.go]: Bad rebase, I have added a comment above these constants. Will address in a separate pull request with real values taken from ERD

File: transport/internal/tlsmux/listener.go (hunk 3 above)
Comment (jronak) [reply transport/internal/tlsmux/listener.go]: Bad rebase again, the errors were logged and metrics were emitted in the handler. This error is used for closing connection

File: transport/internal/tlsmux/listener.go (hunk 2 above)
Comment (jronak) [reply transport/internal/tlsmux/listener.go]: Will address this in a separate PR

File: transport/grpc/options.go (hunk 1 above)
Comment (jronak) [reply transport/grpc/options.go:106]: Added a comment inline with Logger setter:
```
// Logger sets a logger to use for internal logging.
//
// The default is to not write any logs.
```

File: transport/grpc/config.go [hunk 6]
```diff
@@ -285,7 +285,9 @@ func (t *transportSpec) buildTransport(transportConfig *TransportConfig, _ *yarp
 		return nil, err
 	}
 	options = append(options, BackoffStrategy(backoffStrategy))
-	return newTransport(newTransportOptions(options)), nil
+	tr := newTransport(newTransportOptions(options))
+	tr.serviceName = kit.ServiceName()
```
[the remark this answers is not in the document]
Comment (jronak) [reply transport/grpc/config.go]: Good point, I have made it as an option

**Review by gopalchouhan — APPROVED** [review]
Find by me, some comments will be addressed in separate PR.

File: transport/grpc/inbound.go [hunk 7]
```diff
@@ -118,8 +121,24 @@ func (i *Inbound) start() error {
 		grpc.MaxSendMsgSize(i.t.options.serverMaxSendMsgSize),
 	}
 
+	listener := i.listener
+
 	if i.options.creds != nil {
 		serverOptions = append(serverOptions, grpc.Creds(i.options.creds))
+	} else if i.options.tlsMode != yarpctls.Disabled {
+		if i.options.tlsConfig == nil {
+			return errors.New("grpc TLS configuration not provided")
```
Comment (AllenLuUber) [inline transport/grpc/inbound.go:130]: `TLS enabled but configuration not provided`?

File: yarpcconfig/builder.go [hunk 8]
```diff
@@ -326,3 +329,13 @@ func (b *builder) AddStreamOutbound(
 func (b *builder) needTransport(spec *compiledTransportSpec) {
 	b.needTransports[spec.Name] = spec
 }
+
+func tlsModeDecodeHook(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+	var mode yarpctls.Mode
+	if from.Kind() != reflect.String || to != reflect.TypeOf(mode) {
```
Comment (AllenLuUber) [inline yarpcconfig/builder.go:335]: do you want to create a const like what we have in `spec.go` around L247 for this mode?

**Review by AllenLuUber — APPROVED** [review]

File: yarpcconfig/builder.go (hunk 8 above)
Comment (jronak) [reply yarpcconfig/builder.go:335]: Consts in spec.go are for pointer types, but mode here is a non-pointer type so thought of keeping it here

**Pull request merged**